src/core/frames/frame_controller.ts

```diff
@@ -427,17 +427,43 @@ export class FrameController
   }
 
   private handleFrameMissingFromResponse(fetchResponse: FetchResponse) {
-    this.showFrameMissingError()
-    this.throwFrameMissingException(fetchResponse)
+    if (this.responsePermittedToBreakOutOfFrame(fetchResponse)) {
+      this.warnFrameMissingBreakout(fetchResponse)
+      this.visitResponse(fetchResponse.response)
+    } else {
+      this.showFrameMissingError()
+      this.throwFrameMissingException(fetchResponse)
+    }
+  }
+
+  private responsePermittedToBreakOutOfFrame(fetchResponse: FetchResponse) {
+    const allowedPathsTag = this.element.ownerDocument.querySelector<HTMLMetaElement>(
+      `meta[name=turbo-frame-breakout-paths]`
+    )
+    const allowedPaths = allowedPathsTag?.content?.split(/\s+/) || []
+
+    return allowedPaths.includes(fetchResponse.location.pathname)
+  }
+
+  private warnFrameMissingBreakout(fetchResponse: FetchResponse) {
+    console.warn(this.frameMissingMessage(fetchResponse, "Performing a full-page visit."))
   }
 
   private showFrameMissingError() {
     this.element.innerHTML = `<strong class="turbo-frame-error">Content missing</strong>`
   }
```

Comment:
I realise there are some very tiny methods here, but adding them made the logic of `handleFrameMissingFromResponse` read more clearly.

Reply:
I agree. They're worth it as explaining methods, but they also allow higher-level functions to operate at the same level of abstraction. +1